modules/dhcp_common/server.rb

```diff
@@ -105,12 +129,16 @@ def add_record options = {}
       raise(Proxy::DHCP::Error, "DHCP implementation does not support Vendor Options") if vendor_options_included?(options) && !vendor_options_supported?
 
       # try to figure out if we already have this record
-      record = service.find_host_by_ip(subnet.network, options[:ip]) || service.find_host_by_mac(subnet.network, options[:mac])
-      unless record.nil?
+      similar_records = find_similar_records(subnet, options)
```

Comment:
I just realized there is a bug here (always have been) -- find_similar_records only checks reservations, while there might be a conflict with a lease. Providers will catch this, but handling the conflict here would allow for a clearer error message. Again, I think a dedicated PR would be ok for this.